File: 백준/Gold/1931. 회의실 배정/회의실 배정.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        Meeting[] meetings = new Meeting[n];

        StringTokenizer st;
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            meetings[i] = new Meeting(start, end);
        }

        Arrays.sort(meetings);

        int count = 0;
        int end = 0;
        for (int i = 0; i < n; i++) {
            if (meetings[i].start >= end) {
                count++;
                end = meetings[i].end;
            }
        }

        System.out.println(count);
    }

    static class Meeting implements Comparable<Meeting> {
        int start;
        int end;

        public Meeting(int start, int end) {
            this.start = start;
            this.end = end;
        }

        @Override
        public int compareTo(Meeting o) {
            int result = Integer.compare(this.end, o.end);
            if (result == 0) {
                return Integer.compare(this.start, o.start);
            }

            return result;
        }
    }
}

File: 백준/Gold/1931. 회의실 배정/회의실 배정.kt
fun main() {
    val n = readln().toInt()
    val meetings = Array(n) {
        val (start, end) = readln().split(" ").map(String::toInt)
        Meeting(start, end)
    }.sortedWith(
        compareBy<Meeting> { it.end }
            .thenBy { it.start }
    )

    var end = 0
    var count = 0

    for ((start, finish) in meetings) {
        if (start >= end) {
            count++
            end = finish
        }
    }

    print(count)
}

data class Meeting(val start: Int, val end: Int)